app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb

```diff
@@ -128,12 +128,55 @@ def process_collection_item(item, key)
       new_result
     end
 
+    def cross_link_amazon_node(new_result)
+      _, az, uid = URI(new_result[:identity_infra]).path.split('/')
+      az_id = AvailabilityZone.find_by(:name => az).try(:id)
+      ManageIQ::Providers::Amazon::CloudManager::Vm.find_by(
+        :uid_ems              => uid,
+        :availability_zone_id => az_id,
+        :type                 => ManageIQ::Providers::Amazon::CloudManager::Vm.name)
+    end
+
+    def cross_link_google_node(new_result)
+      uri = URI(new_result[:identity_infra])
+      _, az, instance_name = uri.path.split('/')
+      az_id = AvailabilityZone.find_by(:name => az).try(:id)
+      ManageIQ::Providers::Google::CloudManager.find_by_project(uri.host).vms.find_by(
+        :name                 => instance_name,
+        :availability_zone_id => az_id)
+    end
+
+    def cross_link_node(new_result)
+      # Establish a relationship between this node and the vm it is on (if it is in the system)
+      # supported relationships: oVirt, Openstack and VMware.
+      uuid_provider_types = [ManageIQ::Providers::Redhat::InfraManager::Vm.name,
+                             ManageIQ::Providers::Openstack::CloudManager::Vm.name,
+                             ManageIQ::Providers::Vmware::InfraManager::Vm.name]
+
+      # Searching for the underlying instance for Openstack or oVirt.
+      host_instance = Vm.find_by(:uid_ems => new_result[:identity_system].downcase,
+                                 :type    => uuid_provider_types)
+      if !host_instance && new_result[:identity_infra]
+        case URI(new_result[:identity_infra]).scheme
+        when "aws"
+          host_instance = cross_link_amazon_node(new_result)
+        when "gce"
+          host_instance = cross_link_google_node(new_result)
+        end
+      end
```

Comment:
so ```cross_link_amazon_node```, ```cross_link_google_node``` and the logic in line 156-158 are pretty similar. Its worth to think about unifying the functionality here instead. Also, if you prefer you can leave out ```if host_instance``` and do something like: 
```Ruby
new_result[:lives_on_id] = host_instance.try(:id)
new_result[:lives_on_type] = host_instance.try(:type)
```